eclipse-che/che-operator
# feat: configure DWOC from che-operator CR

### What does this PR do?
This PR adds a new reconciler which is in charge of creating and updating a DevWorkspace-Operator configuration (DWOC) in the che cluster namespace. 
The intention is to propagate relevant fields from the che cluster custom resource to the DWOC.

Currently, only the following (storage related) fields are propagated from the che cluster custom resource to the DWOC:
- `checluster.spec.devEnvironments.Storage.pvcStrategy`
- `checluster.spec.devEnvironments.Storage.perUserStrategyPvcConfig.ClaimSize`
- `checluster.spec.devEnvironments.Storage.perUserStrategyPvcConfig.StorageClass`
- `checluster.spec.devEnvironments.Storage.perWorkspaceStrategyPvcConfig.ClaimSize`
- `checluster.spec.devEnvironments.Storage.perWorkspaceStrategyPvcConfig.StorageClass`

The DWOC fields which can receive values from the che cluster are:
- `workspace.storageClassName`
- `workspace.defaultStorageSize.common` 
- `workspace.defaultStorageSize.perWorkspace` 


### Screenshot/screencast of this PR
n/a


### What issues does this PR fix or reference?
Part of eclipse/che/issues/21405


### How to test this PR?
There are a few automated tests that were added.

For manual testing (on minikube):
1. [Build and push the che-operator image to quay](https://github.com/eclipse-che/che-operator#build-and-push-custom-che-operator-image)
2. Deploy che using the custom che-operator image: `chectl server:deploy --installer operator -p minikube --che-operator-image=${IMAGE_REGISTRY_HOST}/${IMAGE_REGISTRY_USER_NAME}/che-operator:next`
3. Once deployed, check to ensure the che-owned DWOC has been created:
```YAML
[aobuchow@fedora ~]$ kubectl describe devworkspaceoperatorconfigs.controller.devfile.io -n eclipse-che
Name:         devworkspace-config
Namespace:    eclipse-che
Labels:       <none>
Annotations:  <none>
API Version:  controller.devfile.io/v1alpha1
Config:
Kind:                    DevWorkspaceOperatorConfig
Metadata:
...
```
4. Edit the che cluster's storage-related fields so that they can be propagated to the DWOC: `kubectl edit checluster eclipse-che -n eclipse-che`
```diff
...
  devEnvironments:
    defaultComponents:
    - container:
        image: quay.io/devfile/universal-developer-image:ubi8-38da5c2
        sourceMapping: /projects
      name: universal-developer-image
    defaultEditor: eclipse/che-theia/latest
    defaultNamespace:
      template: <username>-che
    secondsOfInactivityBeforeIdling: 1800
    secondsOfRunBeforeIdling: -1
    storage:
+     perUserStrategyPvcConfig:
+       claimSize: 15Gi
+       storageClass: testName
      pvcStrategy: per-user
```
5. Ensure the appropriate changes have been made to the DWOC. In this example, the common storage size should be set to `15Gi` and the storage class name should be set to `testName`.
```diff
[aobuchow@fedora ~]$ kubectl describe devworkspaceoperatorconfigs.controller.devfile.io -n eclipse-che
Name:         devworkspace-config
Namespace:    eclipse-che
Labels:       <none>
Annotations:  <none>
API Version:  controller.devfile.io/v1alpha1
Config:
  Workspace:
+   Default Storage Size:
+     Common:            15Gi
+   Storage Class Name:  testName
Kind:                    DevWorkspaceOperatorConfig
Metadata:
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
